### ISSUE TITLE
typecast operator

### ISSUE THREAD

**Issue by muthutt** [tenstorrent/tt-metal#4858]
Parent Issue: #9106

Current status:
- [x] BFLOAT16 -> UINT32 - Completed
- [x] BFLOAT16 -> UINT16 - Completed
- [x] UINT16 -> BFLOAT16 - Completed
- [x] INT32 -> BFLOAT16 - Completed
- [x] BFLOAT16 -> INT32 - Completed

---- BFLOAT16 MNIST & GPT training unblocked -----

- [x] FP32 <-> BFLOAT16 (w/ rounding) - Completed

---- FP32 MNIST & GPT training unblocked -----

- [x] FP32 -> UINT16
- [x] UINT16 -> FP32
- [x] FP32 -> INT32
- [x] INT32 -> FP32
- [x] BFP8 -> UINT16
- [x] UINT16 -> BFP8
- [x] INT32 -> BFP8
- [x] BFP8 -> INT32
- [x] UINT32 -> BFLOAT16
- [x] FP32 -> UINT32
- [x] UINT32 -> FP32
- [x] UINT32 -> BFP8
- [x] BFP8 -> UINT32

**Comment by jliangTT** [comment]
PR reviewed here - https://github.com/tenstorrent-metal/tt-metal/pull/5203

**Comment by jliangTT** [comment]
Recommendation: Can we limit issue of the work to specific conversion between uint32/int16 <> float and have @ttmtrajkovic delivered on this? Further scope for generality support, let's make separate request.

**Comment by razorback3** [comment]
UINT32 -> BFLOAT16: OK
UINT32 -> BFP8: OK
UINT32 -> FP32: Not supported
UINT32 -> FP16: Not supported
BFP8 -> UINT32: OK
BFLOAT16 -> UINT32: OK
FP32 -> UINT32: Not supported
FP16 -> UINT32: Not supported

Is my understanding correct?

**Comment by ttmtrajkovic** [comment]
@razorback3,

I am not sure if any of the conversions work for you yet, they might using the staircase (slow) method. Efficient implementations are not available and can't work at the moment. 
I can work on the following conversions:

UINT32 -> BFLOAT16
UINT16 -> BFLOAT16
BFLOAT16 -> UINT32
BFLOAT16 -> UINT16

UINT16 conversions will be super-fast, using the built-in instruction, while UINT32 conversions will have some custom sfpi code that does exponent extraction, shifting, rounding. Do you need UINT16 conversions?

As a second step, I can provide support for the following:

FP32 -> UINT32/UINT16
UINT16/UINT32 -> FP32

FP16 format is not supported in `ttnn` stack so you can't use it.

Let me know if you have any questions.

Milos

**Comment by razorback3** [comment]
Here is Moreh's priority:
1. UINT32/16 <-> BFLOAT16
2. FP32 <-> BF16 (by the way, is this available?)
3. UINT32/16 <-> FP32
Number 1 is highest.

Thank you :)

**Comment by ttmtrajkovic** [comment]
Item #2 Should already be available if you're ok with just simple truncation (no rounding). Setting up input tensor as fp32 and output tensor as BFLOAT16 in the simple `copy` op should do the work.

milos

**Comment by jliangTT** [comment]
@ttmtrajkovic , can you provide the status for the support for the following? 

UINT32 -> BFLOAT16
UINT16 -> BFLOAT16
BFLOAT16 -> UINT32
BFLOAT16 -> UINT16

**Comment by jliangTT** [comment]
update from speaking to @ttmtrajkovic: 

- there has been priority shuffle so this is going more slowly with best effort. 
- he is still making progress and am planning to work on it today. Will check in weekly to make sure there is forward movement.

**Comment by jliangTT** [comment]
no new status. will check in in a week.

**Comment by davorchap** [comment]
Synced with @razorback3 , these are outstanding

```cpp
FP32 -> UINT32/UINT16
UINT16/UINT32 -> FP32
```

**Comment by ttmtrajkovic** [comment]
Hi @razorback3, @davorchap,
I haven't been able to work on these in the past few weeks, however, those are now reassigned to @rdjogoTT so some progress is being made. Stay tuned for updates

**Comment by davorchap** [comment]
> Hi @razorback3, @davorchap,
> 
> I haven't been able to work on these in the past few weeks, however, those are now reassigned to @rdjogoTT so some progress is being made. Stay tuned for updates

Great, thank you!

**Comment by rdjogoTT** [comment]
Progress update:
- FP32 -> UINT32 LLK implemented, working on enabling the typecast op in order to do final testing
- FP32 -> UINT16 should be straightforward extension of UINT32

**Comment by ayerofieiev-tt** [comment]
Please see an update on #8540 
https://github.com/tenstorrent/tt-metal/issues/8540#issuecomment-2143229766

**Comment by jvasilje** [comment]
@rdjogoTT how is the remaining work going? any updates?

**Comment by rdjogoTT** [comment]
> @rdjogoTT how is the remaining work going? any updates?

FP32->UINT16 is nearing completion, the llk is functional and tested. Just need to add params to the op now to be able to choose which typecast llk to use depending on desired dtypes.

**Comment by ayerofieiev-tt** [comment]
@rdjogoTT completed the typecast for BFLOAT16 --> UINT16

I am updating my PR to properly support this now.

**Comment by rdjogoTT** [comment]
@razorback3 before I start work on UINT32 -> BFLOAT16, would it be sufficient to support INT32 -> BFLOAT16 or do we specifically need UINT32? There is a faster path for implementing INT32.

**Comment by razorback3** [comment]
> @razorback3 before I start work on UINT32 -> BFLOAT16, would it be sufficient to support INT32 -> BFLOAT16 or do we specifically need UINT32? There is a faster path for implementing INT32.

For now, int32 should be affordable. I think there is no use-case that uses full bits of uint32 right now.

**Comment by rdjogoTT** [comment]
> > @razorback3 before I start work on UINT32 -> BFLOAT16, would it be sufficient to support INT32 -> BFLOAT16 or do we specifically need UINT32? There is a faster path for implementing INT32.
> 
> For now, int32 should be affordable. I think there is no use-case that uses full bits of uint32 right now.

@ayerofieiev-tt raised the concern that if we implement INT32->BFLOAT16, the BFLOAT16->UINT32->BFLOAT16 loop would not be possible. Is this loop a requirement or can we just move forward with INT32 for now?

**Comment by rdjogoTT** [comment]
Progress update:

UINT16 -> BFLOAT16 implemented and merged into main.
INT32 -> BFLOAT16 kernel writing underway, should be completed Monday. Opting for INT32 rather than UINT32 due to faster implementation path and no use case that needs full uint32 bits for now.

**Comment by razorback3** [comment]
> > > @razorback3 before I start work on UINT32 -> BFLOAT16, would it be sufficient to support INT32 -> BFLOAT16 or do we specifically need UINT32? There is a faster path for implementing INT32.
> > 
> > 
> > For now, int32 should be affordable. I think there is no use-case that uses full bits of uint32 right now.
> 
> @ayerofieiev-tt raised the concern that if we implement INT32->BFLOAT16, the BFLOAT16->UINT32->BFLOAT16 loop would not be possible. Is this loop a requirement or can we just move forward with INT32 for now?

@rdjogoTT,
I thought if INT32->BFLOAT16 is supported, then BLOAT16->INT32 will be also available.
So, the loop can be BFLOAT16->INT32->BFLOAT16. Would it be possible?

@rdjogoTT , @ayerofieiev-tt 
Can anyone summarize the currently supported conversions? I mean what is supported and what is not supported.

**Comment by rdjogoTT** [comment]
> > > > @razorback3 before I start work on UINT32 -> BFLOAT16, would it be sufficient to support INT32 -> BFLOAT16 or do we specifically need UINT32? There is a faster path for implementing INT32.
> > > 
> > > 
> > > For now, int32 should be affordable. I think there is no use-case that uses full bits of uint32 right now.
> > 
> > 
> > @ayerofieiev-tt raised the concern that if we implement INT32->BFLOAT16, the BFLOAT16->UINT32->BFLOAT16 loop would not be possible. Is this loop a requirement or can we just move forward with INT32 for now?
> 
> @rdjogoTT, I thought if INT32->BFLOAT16 is supported, then BLOAT16->INT32 will be also available. So, the loop can be BFLOAT16->INT32->BFLOAT16. Would it be possible?
> 
> @rdjogoTT , @ayerofieiev-tt Can anyone summarize the currently supported conversions? I mean what is supported and what is not supported.

Ok, I will make sure the loop is supported.

Currently supported conversions:
- BFLOAT16 -> UINT32
- BFLOAT16 -> UINT16
- UINT16 -> BFLOAT16

Not yet supported (next step):
- INT32<->BFLOAT16

**Comment by razorback3** [comment]
To make clear, FP32 <-> BF16 is also currently working with bit truncation. Right?
Is WH supports conversion between them with rounding?

**Comment by rdjogoTT** [comment]
> To make clear, FP32 <-> BF16 is also currently working with bit truncation. Right? Is WH supports conversion between them with rounding?

The answer is yes to both questions. For the second question: this would require unpack to dest support for fp32 to be added as well as a new sfpu kernel.

**Comment by razorback3** [comment]
> > To make clear, FP32 <-> BF16 is also currently working with bit truncation. Right? Is WH supports conversion between them with rounding?
> 
> The answer is yes to both questions. For the second question: this would require unpack to dest support for fp32 to be added as well as a new sfpu kernel.

I see.
Then, I think this will be the remaining request from Moreh to unblock LLM training:
INT32 <-> BFLOAT16
FP32 <-> BFLOAT16

Thanks for your support :)

(cc. @dongjin-na , @namhyeong-kim )

**Comment by rdjogoTT** [comment]
> > > To make clear, FP32 <-> BF16 is also currently working with bit truncation. Right? Is WH supports conversion between them with rounding?
> > 
> > 
> > The answer is yes to both questions. For the second question: this would require unpack to dest support for fp32 to be added as well as a new sfpu kernel.
> 
> I see. Then, I think this will be the remaining request from Moreh to unblock LLM training: INT32 <-> BFLOAT16 FP32 <-> BFLOAT16
> 
> Thanks for your support :)
> 
> (cc. @dongjin-na , @namhyeong-kim )

Could you just please specify if you need FP32 -> BFLOAT16 with rounding or not?

**Comment by razorback3** [comment]
> Could you just please specify if you need FP32 -> BFLOAT16 with rounding or not?

Yeah. I meant with rounding.

**Comment by rdjogoTT** [comment]
Progress update:

- INT32 -> BFLOAT16 kernel complete and tested, just need to run post-commit testing and get approval for PR.
- Next step is BFLOAT16 -> INT32 to complete the loop and then FP32 <-> BFLOAT16 with rounding.

**Comment by rdjogoTT** [comment]
Progress update:

- BFLOAT16 -> INT32 in progress, should be done Wednesday. 
- FP32 <-> BFLOAT16 with rounding to follow

**Comment by sraizada-tt** [comment]
Can we get support for uint16->uint32 too? https://github.com/tenstorrent/tt-metal/issues/9441
I see you have uint16->bfloat16 and bfloat16->uint32 working

**Comment by yieldthought** [comment]
Background on this: we are asking for UINT16->UINT32 because the top-k op outputs UINT16 only but the embedding op accepts UINT32 only and LLMs need to chain these together.

An alternative is UINT32 top-k output or UINT16 embedding input support. But this is a strange thing to be missing either way.

**Comment by ttmtrajkovic** [comment]
@yieldthought, 
UINT32/INT32 for indices of top-k can be done but it will slow as HW has a limitation not be able to efficiently transpose INT32 numbers that are in tiles. UINT16 should be ok for embedding table sizes we have (exception is Llama3 that recently came out) so it was ok to proceed with UINT16 implementation.

I think support to do typacst @sraizada-tt requested should be easy to add, for now, until we pplan for int32 / uint32 for top-k indices.

Milos

**Comment by yieldthought** [comment]
In addition to Llama 3, Grok-1 also uses a 128k embedding size. I would not be surprised - especially give llama3's success - if larger embedding sizes become the norm. UINT16 **is de-facto not ok** for embedding table sizes.

What is our plan to support top-k for modern model architectures?

**Comment by ttmtrajkovic** [comment]
if by modern model architectures you mean embeddings larger than 64k then we will have to implement uint32 as indices and do some work on the existing top-k implementation. it's doable, but it will take some time that we need to scope

**Comment by rdjogoTT** [comment]
Progress update:

FP32 <-> BFLOAT16 merged
FP32 <-> UINT16 next up

**Comment by sangwon-chae** [comment]
@rdjogoTT Thanks for your update. Do you have an expecting date for FP32<->UNIT16?

**Comment by rdjogoTT** [comment]
> @rdjogoTT Thanks for your update. Do you have an expecting date for FP32<->UNIT16?

I should be able to get both directions tested and in by the end of tomorrow.

**Comment by rdjogoTT** [comment]
Progress update:

FP32 <-> UINT16 merged
FP32 <-> INT32 by June 27

**Comment by rdjogoTT** [comment]
@razorback3 With the latest merge I have added support for all typecasts requested except UInt32 cases, for which we said Int32 should be good enough for now. I will continue to implement the remaining few typecasts, but can you confirm that this is no longer blocking other work?

**Comment by razorback3** [comment]
> @razorback3 With the latest merge I have added support for all typecasts requested except UInt32 cases, for which we said Int32 should be good enough for now. I will continue to implement the remaining few typecasts, but can you confirm that this is no longer blocking other work?

Yup, no more blocker for right now. Great work!

**Comment by rdjogoTT** [comment]
@razorback3 All requested typecast variations are now merged, I think this issue can be closed.

**Comment by sraizada-tt** [comment]
Hey! Following up on the support for uint16->uint32. Original issue: https://github.com/tenstorrent/tt-metal/issues/9441

**Comment by mtairum** [comment]
A bit more context:
We require  uint16->uint32 for proper top-1 support on device for LLMs. 
Currently `ttnn.topk` outputs `uint16`, so alternatively, if that  would output to `uint32`, we won't require the typecast.

`ttnn.argmax` does output to `uint32` but that op does not support device_mesh, which is a requirement for larger models on multi-chip.

**Comment by rdjogoTT** [comment]
> Hey! Following up on the support for uint16->uint32. Original issue: #9441

Sorry, I forgot to add that one to the list. I can start on uint16->uint32 today

**Comment by rdjogoTT** [comment]
UINT16 -> UINT32 has been added along with unit tests, this issue should be all done now

**Comment by tt-rkim** [comment]
@sraizada-tt @rdjogoTT 
We are good to close this now?

**Comment by rdjogoTT** [comment]
> @sraizada-tt @rdjogoTT We are good to close this now?

Yes
